<?xml version="1.0" encoding="utf-8"?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from xarm7_robot.urdf.xacro         | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="xarm7">
  <link name="world"/>
  <joint name="world_joint" type="fixed">
    <parent link="world"/>
    <child link="link_base"/>
    <origin rpy="0.0 0.0 0" xyz="-0.24 -0.005 0"/>
  </joint>
  <joint name="table_joint" type="fixed">
    <parent link="world"/>
    <child link="table"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.36 0.0 -0.06"/>
  </joint>
  <link name="table">
    <visual>
      <geometry>
        <mesh filename="package://xarm_description/meshes/optical_table/visual/optical_table.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://xarm_description/meshes/optical_table/visual/optical_table.dae"/>
      </geometry>
    </collision>
  </link>
  <material name="Black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="Red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="White">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="Silver">
    <color rgba="0.753 0.753 0.753 1.0"/>
  </material>
  <link name="link_base">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.021131 -0.0016302 0.056488"/>
      <mass value="0.88556"/>
      <inertia ixx="0.0030595" ixy="0.00012259" ixz="-0.00062705" iyy="0.0037783" iyz="0.00027023" izz="0.0020125"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link_base.STL"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link_base.STL"/>
      </geometry>
    </collision>
  </link>
  <link name="link1">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.0042142 0.02821 -0.0087788"/>
      <mass value="0.42603"/>
      <inertia ixx="0.0014243" ixy="3.9946E-05" ixz="4.7078E-06" iyy="0.00123" iyz="-0.00026073" izz="0.00099229"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link1.dae"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link1.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint1" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 0.267"/>
    <parent link="link_base"/>
    <child link="link1"/>
    <axis xyz="0 0 1"/>
    <limit effort="50" lower="-6.28318530718" upper="6.28318530718" velocity="3.14"/>
    <dynamics damping="10" friction="1"/>
  </joint>
  <link name="link2">
    <inertial>
      <origin rpy="0 0 0" xyz="-3.3178E-05 -0.12849 0.026337"/>
      <mass value="0.56095"/>
      <inertia ixx="0.0031915" ixy="3.0892E-06" ixz="2.8729E-06" iyy="0.0017468" iyz="-0.0010241" izz="0.00235"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link2.dae"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link2.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint2" type="revolute">
    <origin rpy="-1.5708 0 0" xyz="0 0 0"/>
    <parent link="link1"/>
    <child link="link2"/>
    <axis xyz="0 0 1"/>
    <limit effort="50" lower="-2.059" upper="2.0944" velocity="3.14"/>
    <dynamics damping="10" friction="1"/>
  </joint>
  <link name="link3">
    <inertial>
      <origin rpy="0 0 0" xyz="0.04223 -0.023258 -0.0096674"/>
      <mass value="0.44463"/>
      <inertia ixx="0.0011454" ixy="0.0001654" ixz="-0.0001951" iyy="0.0010503" iyz="0.00012027" izz="0.0011083"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link3.dae"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link3.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint3" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 -0.293 0"/>
    <parent link="link2"/>
    <child link="link3"/>
    <axis xyz="0 0 1"/>
    <limit effort="30" lower="-6.28318530718" upper="6.28318530718" velocity="3.14"/>
    <dynamics damping="5" friction="1"/>
  </joint>
  <link name="link4">
    <inertial>
      <origin rpy="0 0 0" xyz="0.067148 -0.10732 0.024479"/>
      <mass value="0.52387"/>
      <inertia ixx="0.0026033" ixy="0.00053379" ixz="0.00036314" iyy="0.0017649" iyz="-0.00083253" izz="0.0022431"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link4.dae"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link4.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint4" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0.0525 0 0"/>
    <parent link="link3"/>
    <child link="link4"/>
    <axis xyz="0 0 1"/>
    <limit effort="30" lower="-0.19198" upper="3.927" velocity="3.14"/>
    <dynamics damping="5" friction="1"/>
  </joint>
  <link name="link5">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.00023397 0.036705 -0.080064"/>
      <mass value="0.18554"/>
      <inertia ixx="0.00099549" ixy="8.9957E-07" ixz="8.5285E-07" iyy="0.0008853" iyz="-0.00025682" izz="0.00035048"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link5.STL"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link5.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint5" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0.0775 -0.3425 0"/>
    <parent link="link4"/>
    <child link="link5"/>
    <axis xyz="0 0 1"/>
    <limit effort="30" lower="-6.28318530718" upper="6.28318530718" velocity="3.14"/>
    <dynamics damping="5" friction="1"/>
  </joint>
  <link name="link6">
    <inertial>
      <origin rpy="0 0 0" xyz="0.058911 0.028469 0.0068428"/>
      <mass value="0.31344"/>
      <inertia ixx="0.00052034" ixy="-0.00014989" ixz="0.00011658" iyy="0.00067968" iyz="6.116E-05" izz="0.00078345"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link6.STL"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link6.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint6" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 0 0"/>
    <parent link="link5"/>
    <child link="link6"/>
    <axis xyz="0 0 1"/>
    <limit effort="20" lower="-1.69297" upper="3.14159265359" velocity="3.14"/>
    <dynamics damping="2" friction="1"/>
  </joint>
  <link name="link7">
    <inertial>
      <origin rpy="0 0 0" xyz="-1.5846E-05 -0.0046377 -0.012705"/>
      <mass value="0.31468"/>
      <inertia ixx="0.00016985" ixy="1.1102E-07" ixz="5.1075E-08" iyy="0.00011924" iyz="-2.1251E-06" izz="0.00026032"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link7.dae"/>
      </geometry>
      <material name="Silver"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://xarm_description/meshes/xarm7/visual/link7.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint7" type="revolute">
    <origin rpy="-1.5708 0 0" xyz="0.076 0.097 0"/>
    <parent link="link6"/>
    <child link="link7"/>
    <axis xyz="0 0 1"/>
    <limit effort="20" lower="-6.28318530718" upper="6.28318530718" velocity="3.14"/>
    <dynamics damping="2" friction="1"/>
  </joint>

  <!--  Begin Robotiq-140-->
  <joint name="robotiq_ur_coupler_robotiq_2f_85_base_joint" type="fixed">
    <origin rpy="0 0 0.7853981633974483" xyz="0 0 0.014"/>
    <parent link="link7"/>
    <child link="robotiq_arg2f_base_link"/>
  </joint>
  <link name="robotiq_arg2f_base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="8.625E-08 -4.6583E-06 0.03145"/>
      <mass value="0.22652"/>
      <inertia ixx="0.00020005" ixy="-4.2442E-10" ixz="-2.9069E-10" iyy="0.00017832" iyz="-3.4402E-08"
               izz="0.00013478"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_base_link.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotiq_2f_140_gripper_visualization/meshes/collision/robotiq_arg2f_base_link.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="grasp_convenient_link"/>
  <joint name="grasp_convenient_joint" type="fixed">
    <origin rpy="0 -1.57079 0" xyz="0 0 0.19"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="grasp_convenient_link"/>
  </joint>
  <link name="left_outer_knuckle">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.000200000000003065 0.0199435877845359 0.0292245259211331"/>
      <mass value="0.00853198276973456"/>
      <inertia ixx="2.89328108496468E-04" ixy="-1.57935047237397E-19" ixz="-1.93980378593255E-19"
               iyy="1.86719750325683E-04" iyz="-1.21858577871576E-04" izz="1.21905238907251E-04"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_outer_knuckle.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.792156862745098 0.819607843137255 0.933333333333333 1"/>
      </material>
    </visual>
    <!--    <collision>-->
    <!--      <origin rpy="0 0 0" xyz="0 0 0"/>-->
    <!--      <geometry>-->
    <!--        <mesh-->
    <!--            filename="package://robotiq_2f_140_gripper_visualization/meshes/collision/robotiq_arg2f_140_outer_knuckle.stl"/>-->
    <!--      </geometry>-->
    <!--    </collision>-->
  </link>
  <link name="left_outer_finger">
    <inertial>
      <origin rpy="0 0 0" xyz="0.00030115855001899 0.0373907951953854 -0.0208027427000385"/>
      <mass value="0.022614240507152"/>
      <inertia ixx="1.52518312458174E-03" ixy="9.76583423954399E-10" ixz="-5.43838577022588E-10"
               iyy="6.17694243867776E-04" iyz="6.78636130740228E-04" izz="1.16494917907219E-03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_outer_finger.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
    <!--    <collision>-->
    <!--      <origin rpy="0 0 0" xyz="0 0 0"/>-->
    <!--      <geometry>-->
    <!--        <mesh-->
    <!--            filename="package://robotiq_2f_140_gripper_visualization/meshes/collision/robotiq_arg2f_140_outer_finger.stl"/>-->
    <!--      </geometry>-->
    <!--    </collision>-->
  </link>
  <link name="left_inner_finger">
    <inertial>
      <origin rpy="0 0 0" xyz="0.000299999999999317 0.0160078233491243 -0.0136945669206257"/>
      <mass value="0.0104003125914103"/>
      <inertia ixx="2.71909453810972E-04" ixy="1.35402465472579E-21" ixz="-7.1817349065269E-22"
               iyy="7.69100314106116E-05" iyz="6.74715432769696E-05" izz="2.30315190420171E-04"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_inner_finger.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>
  <link name="left_inner_finger_pad">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.027 0.065 0.0075"/>
      </geometry>
      <material name="">
        <color rgba="0.9 0.9 0.9 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.03 0.07 0.0075"/>
      </geometry>
      <material name="">
        <color rgba="0.9 0.0 0.0 1"/>
      </material>
    </collision>
  </link>
  <link name="left_inner_knuckle">
    <inertial>
      <origin rpy="0 0 0" xyz="0.000123011831763771 0.0507850843201817 0.00103968640075166"/>
      <mass value="0.0271177346495152"/>
      <inertia ixx="2.61910379223783E-03" ixy="-2.43616858946494E-05" ixz="-6.37789906117123E-09"
               iyy="2.8270243746167E-04" iyz="-5.37200748039765E-05" izz="2.83695868220296E-04"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_inner_knuckle.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>
  <link name="right_outer_knuckle">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.000200000000003065 0.0199435877845359 0.0292245259211331"/>
      <mass value="0.00853198276973456"/>
      <inertia ixx="2.89328108496468E-06" ixy="-1.57935047237397E-19" ixz="-1.93980378593255E-19"
               iyy="1.86719750325683E-06" iyz="-1.21858577871576E-06" izz="1.21905238907251E-06"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_outer_knuckle.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.792156862745098 0.819607843137255 0.933333333333333 1"/>
      </material>
    </visual>
  </link>
  <link name="right_outer_finger">
    <inertial>
      <origin rpy="0 0 0" xyz="0.00030115855001899 0.0373907951953854 -0.0208027427000385"/>
      <mass value="0.022614240507152"/>
      <inertia ixx="1.52518312458174E-03" ixy="9.76583423954399E-10" ixz="-5.43838577022588E-10"
               iyy="6.17694243867776E-03" iyz="6.78636130740228E-04" izz="1.16494917907219E-03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_outer_finger.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>
  <link name="right_inner_finger">
    <inertial>
      <origin rpy="0 0 0" xyz="0.000299999999999317 0.0160078233491243 -0.0136945669206257"/>
      <mass value="0.0104003125914103"/>
      <inertia ixx="2.71909453810972E-03" ixy="1.35402465472579E-21" ixz="-7.1817349065269E-22"
               iyy="7.69100314106116E-04" iyz="6.74715432769696E-04" izz="2.30315190420171E-03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_inner_finger.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>
  <link name="right_inner_finger_pad">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.027 0.065 0.0075"/>
      </geometry>
      <material name="">
        <color rgba="0.9 0.9 0.9 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.03 0.07 0.0075"/>
      </geometry>
      <material name="">
        <color rgba="0.9 0.0 0.0 1"/>
      </material>
    </collision>
  </link>
  <link name="right_inner_knuckle">
    <inertial>
      <origin rpy="0 0 0" xyz="0.000123011831763771 0.0507850843201817 0.00103968640075166"/>
      <mass value="0.0271177346495152"/>
      <inertia ixx="2.61910379223783E-02" ixy="-2.43616858946494E-04" ixz="-6.37789906117123E-06"
               iyy="2.8270243746167E-03" iyz="-5.37200748039765E-04" izz="2.83695868220296E-02"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh
            filename="package://robotiq_2f_140_gripper_visualization/meshes/visual/robotiq_arg2f_140_inner_knuckle.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>

  <joint name="robotiq_2f_85_left_driver_joint" type="prismatic">
    <origin rpy="1.57079 0 0" xyz="0 -0.068 0.177"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="left_inner_finger_pad"/>
    <axis xyz="0 -0.35112344 0.93632918"/>
    <limit effort="1000" lower="-0.06835" upper="0" velocity="2.0"/>
  </joint>
  <joint name="robotiq_2f_85_right_driver_joint" type="prismatic">
    <origin rpy="1.57079 0 -3.14" xyz="0 0.068 0.177"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="right_inner_finger_pad"/>
    <axis xyz="0 -0.35112344 0.93632918"/>
    <limit effort="1000" lower="-0.06835" upper="0" velocity="2.0"/>
  </joint>

  <joint name="robotiq_2f_85_left_driver_mimic_joint" type="revolute">
    <origin rpy="2.29579632679 0 0" xyz="0 -0.030601 0.054905"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="left_outer_knuckle"/>
    <axis xyz="-1 0 0"/>
    <limit effort="1000" lower="0" upper="0.725" velocity="2.0"/>
    <mimic joint="robotiq_2f_85_left_driver_joint" multiplier="-1" offset="0"/>
  </joint>
  <joint name="robotiq_2f_85_left_coupler_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0.01821998610742 0.0260018192872234"/>
    <parent link="left_outer_knuckle"/>
    <child link="left_outer_finger"/>
    <axis xyz="1 0 0"/>
  </joint>
  <joint name="robotiq_2f_85_left_spring_link_joint" type="revolute">
    <origin rpy="2.29579632679 0 0.0" xyz="0 -0.0127 0.06142"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="left_inner_knuckle"/>
    <axis xyz="-1 0 0"/>
    <limit effort="1000" lower="0" upper="0.8757" velocity="2.0"/>
    <mimic joint="robotiq_2f_85_left_driver_joint" multiplier="-1" offset="0"/>
  </joint>
  <joint name="robotiq_2f_85_left_pad_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 -0.0457554015893473 0.0272203446692936"/>
    <child link="left_inner_finger"/>
    <parent link="left_inner_finger_pad"/>
    <axis xyz="0 0 1"/>
  </joint>

  <joint name="robotiq_2f_85_right_driver_mimic_joint" type="revolute">
    <origin rpy="2.29579632679 0 3.14159265359" xyz="0 0.030601 0.054905"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="right_outer_knuckle"/>
    <axis xyz="-1 0 0"/>
    <limit effort="1000" lower="0.0" upper="0.725" velocity="2.0"/>
    <mimic joint="robotiq_2f_85_left_driver_joint" multiplier="-1" offset="0"/>
  </joint>
  <joint name="robotiq_2f_85_right_coupler_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0.01821998610742 0.0260018192872234"/>
    <parent link="right_outer_knuckle"/>
    <child link="right_outer_finger"/>
    <axis xyz="1 0 0"/>
  </joint>
  <joint name="robotiq_2f_85_right_spring_link_joint" type="revolute">
    <origin rpy="2.29579632679 0 -3.14159265359" xyz="0 0.0127 0.06142"/>
    <parent link="robotiq_arg2f_base_link"/>
    <child link="right_inner_knuckle"/>
    <axis xyz="-1 0 0"/>
    <limit effort="1000" lower="0" upper="0.8757" velocity="2.0"/>
    <mimic joint="robotiq_2f_85_left_driver_joint" multiplier="-1" offset="0"/>
  </joint>
  <joint name="robotiq_2f_85_right_pad_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 -0.0457554015893473 0.0272203446692936"/>
    <child link="right_inner_finger"/>
    <parent link="right_inner_finger_pad"/>
    <axis xyz="0 0 1"/>
  </joint>

  <material name="aluminum">
    <color rgba="0.5 0.5 0.5 1"/>
  </material>
  <material name="plastic">
    <color rgba="0.1 0.1 0.1 1"/>
  </material>
  <joint name="camera_base_joint" type="fixed">
    <origin rpy="0.11337993967393135 0.970870739452661 -2.746000975910688" xyz="0.95055202 0.13056585 0.64754431"/>
    <parent link="link_base"/>
    <child link="camera_base_color_frame"/>
  </joint>
  <link name="camera_base_color_frame"/>

  <joint name="camera_base_color_joint" type="fixed">
    <origin xyz="0 -0.015 0" rpy="0.001 0.005 -0.012"/>
    <child link="camera_base_link"/>
    <parent link="camera_base_color_frame"/>
  </joint>
  <link name="camera_base_link">
    <visual>
      <origin rpy="1.5707963267948966 0 1.5707963267948966" xyz="0.00987 -0.02 -0.02"/>
      <geometry>
        <mesh filename="package://realsense2_description/meshes/d415.stl"/>
      </geometry>
      <material name="aluminum"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 -0.02 0"/>
      <geometry>
        <box size="0.02005 0.099 0.023"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.564"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.003881243" ixy="0.0" ixz="0.0" iyy="0.000498940" iyz="0.0" izz="0.003879257"/>
    </inertial>
  </link>
  <joint name="camera_base_color_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="-1.571 -0.000 -1.571"/>
    <parent link="camera_base_color_frame"/>
    <child link="camera_base_color_optical_frame"/>
  </joint>
  <link name="camera_base_color_optical_frame"/>

  <gazebo reference="camera_base_color_frame">
    <sensor type="depth" name="camera_base">
      <camera>
        <horizontal_fov>1.2108</horizontal_fov>
        <image>
          <width>1920</width>
          <height>1080</height>
        </image>
        <clip>
          <near>0.01</near>
          <far>100</far>
        </clip>
      </camera>
    </sensor>
  </gazebo>
  <!-- Only used when processing real point cloud since it requires camera_base_depth_optical_frame -->
  <joint name="camera_base_depth_optical_joint" type="fixed">
    <origin xyz="-0.000 -0.015 -0.000" rpy="-1.576 0.001 -1.583"/>
    <parent link="camera_base_color_frame"/>
    <child link="camera_base_depth_optical_frame"/>
  </joint>
  <link name="camera_base_depth_optical_frame"/>

  <joint name="camera_hand_joint" type="fixed">
    <origin rpy="3.007427956290561 -1.553556250188954 0.8798708316363683" xyz="0.07573088 0.0195131  0.01554786"/>
    <parent link="link7"/>
    <child link="camera_hand_color_frame"/>
  </joint>
  <link name="camera_hand_link">
    <visual>
      <origin rpy="1.5707963267948966 0 1.5707963267948966" xyz="0.00987 -0.02 -0.02"/>
      <geometry>
        <mesh filename="package://realsense2_description/meshes/d415.stl"/>
      </geometry>
      <material name="aluminum"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 -0.02 0"/>
      <geometry>
        <box size="0.02005 0.099 0.023"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.564"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.003881243" ixy="0.0" ixz="0.0" iyy="0.000498940" iyz="0.0" izz="0.003879257"/>
    </inertial>
  </link>

  <!-- camera color joints and links -->
  <joint name="camera_hand_color_joint" type="fixed">
    <origin xyz="0 -0.015 0" rpy="0 0 0"/>
    <parent link="camera_hand_color_frame"/>
    <child link="camera_hand_link"/>
  </joint>
  <link name="camera_hand_color_frame"/>

  <joint name="camera_hand_color_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="-1.571 -0.000 -1.571"/>
    <parent link="camera_hand_color_frame"/>
    <child link="camera_hand_color_optical_frame"/>
  </joint>
  <link name="camera_hand_color_optical_frame"/>
  <gazebo reference="camera_hand_color_frame">
    <sensor type="depth" name="camera_hand">
      <camera>
        <horizontal_fov>1.2108</horizontal_fov>
        <image>
          <width>1920</width>
          <height>1080</height>
        </image>
        <clip>
          <near>0.01</near>
          <far>100</far>
        </clip>
      </camera>
    </sensor>
  </gazebo>
  <!-- Only used when processing real point cloud since it requires camera_hand_depth_optical_frame -->
  <joint name="camera_hand_depth_optical_joint" type="fixed">
    <origin xyz="-0.000 -0.015 -0.000" rpy="-1.573 0.003 -1.576"/>
    <parent link="camera_hand_color_frame"/>
    <child link="camera_hand_depth_optical_frame"/>
  </joint>
  <link name="camera_hand_depth_optical_frame"/>
</robot>
